contrib/check-semver.sh

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env bash
+#
+# Checks the public API of crates, exits with non-zero if there are currently
+# changes to the public API not already committed to in the various api/*.txt
+# files.
+
+set -euo pipefail
+
+NIGHTLY=$(cat nightly-version)
+# Our docs have broken intra doc links if all features are not enabled.
+RUSTDOCFLAGS="-Z unstable-options --document-private-items --document-hidden-items --output-format=json --cap-lints=allow"
+CURRENT_COMMIT=$(git rev-parse HEAD)
+MASTER_COMMIT=$(git rev-parse master)
+
+main() {
+    need_nightly
+
+    # on current commit
+    generate_json_files "bitcoin" "current"
+    generate_json_files "base58ck" "current"
+    generate_json_files "hashes" "current"
+    generate_json_files "units" "current"
+    generate_json_files "io" "current"
+
+    # switch to master
+    git checkout "$MASTER_COMMIT"
+
+    generate_json_files "bitcoin" "master"
+    generate_json_files "base58" "master"
+    generate_json_files "hashes" "master"
+    generate_json_files "units" "master"
+    generate_json_files "io" "master"
+
+    # Check for API semver breaks
+    run_cargo_semver_check "bitcoin"
+    run_cargo_semver_check "base58ck"
+    run_cargo_semver_check "hashes"
+    run_cargo_semver_check "units"
+    run_cargo_semver_check "io"
+
+    #
+    check_for_changes
+    
+    # switch back again
+    git checkout "$CURRENT_COMMIT"
+}
+
+# Run cargo doc
+run_cargo_doc() {
+    RUSTDOCFLAGS="$RUSTDOCFLAGS" cargo +"$NIGHTLY" doc "$@"
+}
+
+# Run cargo semver-check
+run_cargo_semver_check() {
+    local crate="$1"
+    # FIXME: how do I pipe the output of `cargo semver-checks` to a file?
+    cargo +"$NIGHTLY" semver-checks --json  --baseline-rustdoc "$crate-master-no-features.json" --current-rustdoc "$crate-current-all-features.json" > "$crate-semver.txt"
```

Comment:
@obi1kenobi how do I save the `cargo semver-checks` output to a file?

Reply:
Append `2>&1` to the command. Right now you are sending stdout to the file, but stderr still gets printed normally. If you do `>file.txt 2>&1` then that sends both stdout and stderr to the file. The order of these two commands matters, so make sure `2>&1` is second.

Reply:
We don't have a structured output format for the cargo-semver-checks findings, but I'm open to adding one. We just have to work out a good name for the CLI flag, and how to version it and keep it stable across versions.

Reply:
For now I just want to `grep` a fail in the `-v` (verbose) output and if positive, add a label to the PR (and maybe a comment as well?).